sw/device/lib/dif/dif_kmac.h

```diff
@@ -0,0 +1,797 @@
+   */
+  kDifKmacIrqDone = 0,
+
+  /**
+   * Message FIFO empty condition.
+   *
+   * Associated with the `kmac.INTR_STATE.fifo_empty` hardware interrupt.
+   */
+  kDifKmacIrqFifoEmpty,
+
+  /**
+   * KMAC/SHA3 error occurred.
+   *
+   * Associated with the `kmac.INTR_STATE.kmac_err` hardware interrupt.
+   */
+  kDifKmacIrqError,
+} dif_kmac_irq_t;
+
+/**
+ * Supported SHA-3 modes of operation.
+ */
+typedef enum dif_kmac_mode_sha3 {
+  /** SHA-3 with 224 bit strength. */
+  kDifKmacModeSha3L224,
+  /** SHA-3 with 256 bit strength. */
+  kDifKmacModeSha3L256,
+  /** SHA-3 with 384 bit strength. */
+  kDifKmacModeSha3L384,
+  /** SHA-3 with 512 bit strength. */
+  kDifKmacModeSha3L512,
+} dif_kmac_mode_sha3_t;
+
+/**
+ * Supported SHAKE modes of operation.
+ */
+typedef enum dif_kmac_mode_shake {
+  /** SHAKE with 128 bit strength. */
+  kDifKmacModeShakeL128,
+  /** SHAKE with 256 bit strength. */
+  kDifKmacModeShakeL256,
+} dif_kmac_mode_shake_t;
+
+/**
+ * Supported cSHAKE modes of operation.
+ */
+typedef enum dif_kmac_mode_cshake {
+  /** cSHAKE with 128 bit strength. */
+  kDifKmacModeCshakeL128,
+  /** cSHAKE with 256 bit strength. */
+  kDifKmacModeCshakeL256,
+} dif_kmac_mode_cshake_t;
+
+/**
+ * Supported KMAC modes of operation.
+ */
+typedef enum dif_kmac_mode_kmac {
+  /** KMAC with 128 bit strength. */
+  kDifKmacModeKmacL128,
+  /** KMAC with 256 bit strength. */
+  kDifKmacModeKmacL256,
+} dif_kmac_mode_kmac_t;
+
+/**
+ * Maximum lengths supported by the KMAC unit.
+ */
+enum {
+  kDifKmacMaxKeyBytes = 512 / 8,
+
+  /**
+   * The maximum length in bytes of a customization string (S) before it has
+   * been encoded.
+   */
+  kDifKmacMaxCustomizationStringLen = 32,
+
+  /**
+   * The maximum number of bytes required to encode the length of the
+   * customization string.
+   *
+   * Assumes maximum customization string length of 32 bytes (256 bits).
+   */
+  kDifKmacMaxCustomizationStringOverhead = 3,
+
+  /**
+   * The maximum length in bytes of a function name (N) before it has been
+   * encoded.
+   */
+  kDifKmacMaxFunctionNameLen = 4,
+
+  /**
+   * The maximum number of bytes required to encode the length of the function
+   * name.
+   *
+   * Assumes maximum function name length of 4 bytes (32 bits).
+   */
+  kDifKmacMaxFunctionNameOverhead = 2,
+};
+
+/**
+ * Key type.
+ */
+typedef enum dif_kmac_key_type {
+  /** Software provided 128 bit key. */
+  kDifKmacKeyTypeL128,
+  /** Software provided 192 bit key. */
+  kDifKmacKeyTypeL192,
+  /** Software provided 256 bit key. */
+  kDifKmacKeyTypeL256,
+  /** Software provided 384 bit key. */
+  kDifKmacKeyTypeL384,
+  /** Software provided 512 bit key. */
+  kDifKmacKeyTypeL512,
+} dif_kmac_key_type_t;
+
+/**
+ * A key for KMAC operations.
+ *
+ * The key is provided in two parts, `share0` and `share1`. These are
+ * combined using a bitwise XOR operation in the KMAC unit to produce the real
+ * key.
+ *
+ * The key shares are encoded in little endian byte order.
```

Comment:
> little endian byte order. 

nit: cross-link to input/output endianess parameters.

Reply:
Mentioned that the endianness here is fixed as opposed to the message/output endianness (though I'm still not sure that 'big-endian' mode is actually useful for those - personally I'd just let software do that).